### PR TITLE
Asset Prices

### PR DESCRIPTION
**Pull request by walmat**
Fixes APP-1343
Fixes APP-1457

## What changed (plus any additional context for devs)
Added asset price fetching and some react-query external token handling.

### CODE REVIEW

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts [hunk 1]
```diff
@@ -196,20 +194,108 @@ export function useSwapInputsController({
...
+      const currency = store.getState().settings.nativeCurrency;
+
+      try {
+        const tokenData = await fetchExternalToken({
+          address,
+          network,
+          currency,
+        });
+
+        queryClient.setQueryData(externalTokenQueryKey({ address, network, currency }), tokenData);
```
Comment (benisgold) [inline src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:263]: do we need to worry about setting a stale time or anything here? i have no idea just curious

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts [hunk 2]
```diff
@@ -196,20 +194,108 @@ export function useSwapInputsController({
...
+    const updateNativePriceForAsset = ({ price, type }: { price: number; type: string }) => {
+      'worklet';
+      if (type === 'inputAsset') {
+        internalSelectedInputAsset.modify(prev => ({ ...prev, nativePrice: price }));
+      } else if (type === 'outputAsset') {
+        internalSelectedOutputAsset.modify(prev => ({ ...prev, nativePrice: price }));
+      }
+    };
+
+    const getAssetNativePrice = async ({ asset, type }: { asset: ExtendedAnimatedAssetWithColors | null; type: string }) => {
```
Comment (benisgold) [inline src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:249]: maybe slap this bad boy in a useCallback (and others in this hook)

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts [hunk 3]
```diff
@@ -196,20 +194,108 @@ export function useSwapInputsController({
...
+
+        if (tokenData?.price.value) {
+          runOnUI(updateNativePriceForAsset)({
+            price: tokenData.price.value,
+            type,
+          });
+        }
+      } catch (error) {
+        logger.error(new RainbowError('[useSwapInputsController]: get asset prices failed'));
+        const cachedData = await queryClient.getQueryData<ReturnType<typeof fetchExternalToken>>(
```
Comment (benisgold) [inline src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:273]: getQueryData is synchronous

File: src/__swaps__/screens/Swap/hooks/useSwapNavigation.ts [hunk 4]
```diff
@@ -66,7 +66,7 @@ export function useSwapNavigation({
     'worklet';
```
Comment (benisgold) [inline src/__swaps__/screens/Swap/hooks/useSwapNavigation.ts:66]: worklets are automatically memoized under the hood so useCallback is unnecessary fyi

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts (hunk 1 above)
Comment (walmat) [reply src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:263]: i think ideally we'd want to handle it, but for right now it's probably fine. The cache gets updated on every successful fetch, cache is only used as a fallback.

Comment (walmat) [reply src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:263]: actually thinking twice about it, went ahead and implemented an invalidation strategy

**Review by benisgold — APPROVED** [review]

**Pull request merged**